### PR TITLE
Performance optimization 2

### PR DESCRIPTION
**Pull request by halil81**
This PR adds `caching` behavior to some methods that return a _lazy_ `Iterable`.

### CODE REVIEW

File: lib/src/serializers/json_api.dart [hunk 1]
```diff
@@ -249,7 +272,17 @@ class JsonApiDocument {
           _isAttributeError(error, attributeName) && _hasErrorDetail(error))
       : false;
 
-  List<String> errorsFor(String attributeName) => List<String>.from(errors
+  List<String> errorsFor(String attributeName) {
+    var key = 'errorsFor:${attributeName}';
+    List<String>? ids = _idsCache[key];
+    if (ids == null) {
+      ids = _errorsFor(attributeName);
```
Comment (algodave) [inline lib/src/serializers/json_api.dart:279]: @halil81 The `_errorsFor` method returns a `List` of error messages, not IDs. I agree to reuse the same cache, however, let's rename it from `_idsCache` to `_stringsCache`.

**Review by algodave — CHANGES_REQUESTED** [review]

File: lib/src/serializers/json_api.dart [hunk 2]
```diff
@@ -74,6 +74,9 @@ class JsonApiDocument {
   Map<String, dynamic> relationships;
   List<dynamic> included;
   List<dynamic> errors;
+  Map<String, List<String>> _idsCache = Map<String, List<String>>();
+  Map<String, List<JsonApiDocument>> _docsCache =
```
Comment (algodave) [inline lib/src/serializers/json_api.dart:78]: @halil81 This is unused, as reported by Dart analyzer

Comment (halil81) [reply lib/src/serializers/json_api.dart]: @algodave I removed _docsCache. Thanks!

File: lib/src/serializers/json_api.dart (hunk 1 above)
Comment (halil81) [reply lib/src/serializers/json_api.dart]: @algodave I fixed variable names. Thanks!

**Comment by algodave** [comment]
@halil81 We need to avoid all that repetitive code that performs "read from cache or compute and load into cache". Following is my proposal.

```dart

// We can define a simple private class like this:

class _Cache<T> {
  Map<String, T> _map = Map<String, T>();

  T readOrLoad({
    required String key,
    required T Function() loader,
  }) =>
      _map[key] ?? _load(key, loader);

  T _load(String key, T Function() loader) {
    T value = loader();
    _map[key] = value;
    return value;
  }
}

// We would then redefine ids_cache like this:

_Cache<List<String>> _idsCache = _Cache<List<String>>();

// Methods could then become as compact as:

  List<String> includedIdsFor(String relationshipName, String modelType) =>
      _idsCache.readOrLoad(
        key: 'includedIdsFor:${relationshipName}',
        loader: () => _includedIdsFor(relationshipName, modelType),
      );

```

**Pull request merged**